File: renderer/src/legacyTelemetry.ts
import {createHmac} from "node:crypto";

const callerIdPattern = /^[A-Za-z0-9][A-Za-z0-9._:-]{0,63}$/;

type LegacyRoute = "/healthz" | "/render" | "/media/{path}";

const routeAction: Record<LegacyRoute, string> = {
  "/healthz": "health_probe",
  "/render": "legacy_render",
  "/media/{path}": "legacy_media_read",
};

const resolveRoute = (path: string): LegacyRoute | null => {
  if (path === "/healthz" || path === "/render") return path;
  if (path.startsWith("/media/")) return "/media/{path}";
  return null;
};

const safeClaimedCaller = (value: string | undefined): string | null => {
  if (value === undefined) return null;
  const candidate = value.trim();
  return callerIdPattern.test(candidate) ? candidate : "invalid";
};

const fingerprint = (value: string, salt: string | undefined): string | null => {
  if (!salt) return null;
  return `hmac-sha256:${createHmac("sha256", salt).update(value).digest("hex").slice(0, 24)}`;
};

export type LegacyTelemetryInput = {
  path: string;
  method: string;
  statusCode: number;
  peerAddress?: string;
  claimedCallerId?: string;
  userAgent?: string;
};

export const createLegacyTelemetry = (
  salt: string | undefined,
  emit: (event: Record<string, unknown>) => void = (event) => console.log(JSON.stringify(event)),
) => {
  const routeCounts = new Map<LegacyRoute, number>();
  let deprecatedAttemptCount = 0;

  return (input: LegacyTelemetryInput): Record<string, unknown> | null => {
    const route = resolveRoute(input.path);
    if (route === null) return null;
    const action = routeAction[route];
    const deprecatedAttempt = action !== "health_probe";
    const routeRequestCount = (routeCounts.get(route) ?? 0) + 1;
    routeCounts.set(route, routeRequestCount);
    if (deprecatedAttempt) deprecatedAttemptCount += 1;

    const peer = input.peerAddress ?? "unavailable";
    const agent = input.userAgent ?? "unavailable";
    const event: Record<string, unknown> = {
      event: "legacy_route_access",
      service: "video-factory-v1-renderer",
      route,
      method: input.method.toUpperCase(),
      status_code: input.statusCode,
      action,
      deprecated_attempt: deprecatedAttempt,
      claimed_caller_id: safeClaimedCaller(input.claimedCallerId),
      source_fingerprint: fingerprint(peer, salt),
      client_fingerprint: fingerprint(`${peer}\n${agent}`, salt),
      identity_ready: Boolean(salt),
      route_request_count: routeRequestCount,
      deprecated_attempt_count: deprecatedAttemptCount,
      payload_logged: false,
      raw_network_identity_logged: false,
    };
    emit(event);
    return event;
  };
};

File: apps/api/app/legacy_telemetry.py
from __future__ import annotations

import hashlib
import hmac
import json
import logging
import os
import re
from collections import Counter
from datetime import datetime, timezone
from pathlib import Path
from threading import Lock
from typing import Any
from uuid import uuid4


# Uvicorn installs an INFO handler on the ``uvicorn`` logger hierarchy before
# application lifespan starts. Keep telemetry in that hierarchy so structured
# events have a production output path without changing global/root logging or
# adding a second handler that could duplicate records.
LOGGER_NAME = "uvicorn.error.npd.legacy_telemetry"
LOGGER = logging.getLogger(LOGGER_NAME)
CALLER_ID = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._:-]{0,63}$")
ROUTE_ACTIONS = {
    "/healthz": "health_probe",
    "/readyz": "health_probe",
    "/api/v1/video-jobs": "legacy_write",
    "/api/v1/video-jobs/{job_id}": "legacy_read",
    "/api/v1/video-jobs/{job_id}/artifacts/{artifact_name}": "legacy_artifact_read",
}
MINIMUM_SALT_BYTES = 32


def _safe_claimed_caller(value: str | None) -> str | None:
    if value is None:
        return None
    value = value.strip()
    return value if CALLER_ID.fullmatch(value) else "invalid"


def _fingerprint(value: str, salt: bytes | None) -> str | None:
    if not salt:
        return None
    digest = hmac.new(salt, value.encode("utf-8", errors="replace"), hashlib.sha256).hexdigest()
    return f"hmac-sha256:{digest[:24]}"


def _salt_from_environment() -> str | None:
    direct = os.getenv("LEGACY_TELEMETRY_SALT", "").strip()
    file_value = os.getenv("LEGACY_TELEMETRY_SALT_FILE", "").strip()
    if direct and file_value:
        raise RuntimeError(
            "configure only one of LEGACY_TELEMETRY_SALT or LEGACY_TELEMETRY_SALT_FILE"
        )
    if file_value:
        path = Path(file_value)
        if not path.is_absolute():
            raise RuntimeError("LEGACY_TELEMETRY_SALT_FILE must be an absolute path")
        try:
            direct = path.read_text(encoding="utf-8").strip()
        except OSError as exc:
            raise RuntimeError("LEGACY_TELEMETRY_SALT_FILE could not be read") from exc
        if not direct:
            raise RuntimeError("LEGACY_TELEMETRY_SALT_FILE is empty")
    if direct and len(direct.encode("utf-8")) < MINIMUM_SALT_BYTES:
        raise RuntimeError(f"legacy telemetry salt must contain at least {MINIMUM_SALT_BYTES} bytes")
    return direct or None


class LegacyTelemetry:
    """Emit identity-safe route events without storing payloads, IDs, URLs, IPs or user agents."""

    def __init__(self, *, salt: str | None, logger: logging.Logger = LOGGER):
        self._salt = salt.encode("utf-8") if salt else None
        self._logger = logger
        self._process_instance_id = str(uuid4())
        self._route_counts: Counter[str] = Counter()
        self._deprecated_attempt_count = 0
        self._lock = Lock()

    @classmethod
    def from_environment(cls) -> "LegacyTelemetry":
        return cls(salt=_salt_from_environment())

    @property
    def identity_ready(self) -> bool:
        return self._salt is not None

    def record(
        self,
        *,
        route: str | None,
        method: str,
        status_code: int,
        peer_host: str | None,
        claimed_caller_id: str | None,
        user_agent: str | None,
    ) -> dict[str, Any] | None:
        if route not in ROUTE_ACTIONS:
            return None
        action = ROUTE_ACTIONS[route]
        deprecated_attempt = action != "health_probe"
        with self._lock:
            self._route_counts[route] += 1
            route_count = self._route_counts[route]
            if deprecated_attempt:
                self._deprecated_attempt_count += 1
            deprecated_count = self._deprecated_attempt_count

        peer = peer_host or "unavailable"
        agent = user_agent or "unavailable"
        event: dict[str, Any] = {
            "event": "legacy_route_access",
            "observed_at": datetime.now(timezone.utc).isoformat(),
            "process_instance_id": self._process_instance_id,
            "service": "video-factory-v1-api",
            "route": route,
            "method": method.upper(),
            "status_code": int(status_code),
            "action": action,
            "deprecated_attempt": deprecated_attempt,
            "claimed_caller_id": _safe_claimed_caller(claimed_caller_id),
            "source_fingerprint": _fingerprint(peer, self._salt),
            "client_fingerprint": _fingerprint(f"{peer}\n{agent}", self._salt),
            "identity_ready": self.identity_ready,
            "route_request_count": route_count,
            "deprecated_attempt_count": deprecated_count,
            "payload_logged": False,
            "raw_network_identity_logged": False,
        }
        self._logger.info(json.dumps(event, ensure_ascii=False, separators=(",", ":")))
        return event
